openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,634 @@
+    /// previous node in the tree at the same position is replaced upon merging
+    /// the diff.
+    ///
+    /// Returns an error if the given node index is larger than the current size
+    /// of the diff.
+    fn replace_node(&mut self, node_index: NodeIndex, node: T) -> Result<(), ABinaryTreeDiffError> {
+        // Check that we're not out of bounds.
+        self.out_of_bounds(node_index)?;
+        self.diff.insert(node_index, node);
+        Ok(())
+    }
+
+    /// Removes a node from the right edge of the diff, thus decreasing the size
+    /// of the diff by one. Throws an error if this would make the diff too
+    /// small (i.e. < 1 node).
+    fn remove_node(&mut self) -> Result<(), ABinaryTreeDiffError> {
+        // First make sure that the tree isn't getting too small.
+        if self.size() <= 1 {
+            return Err(ABinaryTreeDiffError::TreeTooSmall);
+        }
+        // Then check if the tree was extended before. If so, just remove the
+        // last node from the diff.
+        if self.size() > self.original_tree.size()? {
+            let removed = self.diff.remove(&(self.size() - 1));
+            // There should be a node here to remove.
+            debug_assert!(removed.is_some());
+        } else {
+            // If that is not the case, either the tree is of the same length as
+            // the diff, or the diff is already smaller. In both cases, we check
+            // if there is a node at the right edge to remove from the diff.
+            self.diff.remove(&(self.size() - 1));
+            // Regardless of the result, we decrease the size to signal that a
+            // node was removed from the diff.
+        }
+        self.size -= 1;
+        Ok(())
+    }
+
+    // Index checking
+
+    /// This is a helper function to check if a given leaf is outside of the
+    /// bounds of the tree.
+    ///
+    /// Returns an error if the given leaf is out of bounds. Otherwise returns
+    /// nothing.
+    fn out_of_bounds(&self, node_index: NodeIndex) -> Result<(), ABinaryTreeDiffError> {
+        if node_index >= self.size() {
+            return Err(ABinaryTreeDiffError::OutOfBounds);
+        }
+        Ok(())
+    }
+
+    /// This is a helper function to check the validity of two leaf indices for
+    /// use in the subtree root functions.
+    ///
+    /// Returns an error if the two given leaf indices are the same or if one of
+    /// the leaf indices doesn't correspond to a leaf in the diff. Otherwise
+    /// returns nothing.
+    fn leaf_pair_check(
+        &self,
+        leaf_index_1: LeafIndex,
+        leaf_index_2: LeafIndex,
+    ) -> Result<(), ABinaryTreeDiffError> {
+        let node_index_1 = to_node_index(leaf_index_1);
+        let node_index_2 = to_node_index(leaf_index_2);
+
+        self.out_of_bounds(node_index_1)?;
+        self.out_of_bounds(node_index_2)?;
+
+        if leaf_index_1 == leaf_index_2 {
```

Comment:
Shouldn't this check come first?

Reply:
If you move it all the way up, you don't need to call functions if there's an error. The earlier you can return, the better.